text/0126-introduce-xcq.md

```diff
@@ -0,0 +1,342 @@
+
+- A hash-based extension id generation mechanism for addressing and versioning . The hash value derives from the extension name and its method sets. Any update to an extension is treated as a new extension.
+
+- `decl_extension` macro: Defines an extension as a Rust trait with optional associated types.
+
+**Example usage**:
+```rust
+use xcq_extension::decl_extension;
+
+pub trait Config {
+    type AssetId: Codec;
+    type AccountId: Codec;
+    type Balance: Codec;
+}
+decl_extension! {
+    pub trait ExtensionFungibles {
+        type Config: Config;
+        fn total_supply(asset: <Self::Config as Config>::AssetId) -> <Self::Config as Config>::Balance;
+        fn balance(asset: <Self::Config as Config>::AssetId, who: <Self::Config as Config>::AccountId) -> <Self::Config as Config>::Balance;
+    }
+}
+```
+
+- `impl_extensions` macro: Generates extension implementations and extension-level metadata.
+
+**Example Usage**:
+
+```rust
+// ExtensionImpl is an aggregate struct to impl different extensions
+struct ExtensionImpl;
+impl extension_fungibles::Config for ExtensionImpl {
+    type AssetId = u32;
+    type AccountId = [u8; 32];
+    type Balance = u64;
+}
+impl_extensions! {
+    impl extension_core::ExtensionCore for ExtensionImpl {
+        type Config = ExtensionImpl;
+        fn has_extension(id: <Self::Config as extension_core::Config>::ExtensionId) -> bool {
+            matches!(id, 0 | 1)
+        }
+    }
+
+    impl extension_fungibles::ExtensionFungibles for ExtensionImpl {
+        type Config = ExtensionImpl;
+        #[allow(unused_variables)]
+        fn total_supply(asset: <Self::Config as extension_fungibles::Config>::AssetId) -> <Self::Config as extension_fungibles::Config>::Balance {
+            200
+        }
+        #[allow(unused_variables)]
+        fn balance(asset: <Self::Config as extension_fungibles::Config>::AssetId, who: <Self::Config as extension_fungibles::Config>::AccountId) -> <Self::Config as extension_fungibles::Config>::Balance {
+            100
+        }
+    }
+}
+```
+
+- `ExtensionExecutor`: Connects extension implementations and `xcq-executor`.
+All methods of all extensions that a chain supports are amalgamated into a single `host_call` entry.
+Then this entry is registered as a typed function entry in PolkaVM Linker within the `xcq-executor`.
+Given the extension ID and call data encoded in SCALE format, call requests from the guest XCQ program are dispatched to corresponding extensions:
+```rust
+linker
+    .define_typed(
+        "host_call",
+        move |caller: Caller<'_, Self::UserData>,
+              extension_id: u64,
+              call_ptr: u32,
+              call_len: u32|
+              -> Result<u64, ExtensionError> {
+                  ...
+              });
+```
+- `PermController`: Filters guest XCQ program calling requests, useful for host chains to disable some queries by filtering invoking sources.
+```rust
+pub trait PermController {
```

Comment:
Maybe `Authorizer` is a better name? Or `PermissionsController` (not abbreviated)